client/src/App.test.tsx

```diff
@@ -1,15 +1,177 @@
 import React from 'react'
-import { render, screen } from '@testing-library/react'
+import { screen, render } from '@testing-library/react'
+import { MemoryRouter } from 'react-router-dom'
 import App from './App'
-import * as utilities from './components/utilities'
+import { withMockFetch, renderWithRouter } from './components/testUtilities'
+import { IUserMeta } from './types'
+import { dummyBoards } from './components/DataEntry/_mocks'
+import {
+  // jaApiCalls,
+  aaApiCalls,
+} from './components/MultiJurisdictionAudit/_mocks'
+// import { manifestMocks } from './components/MultiJurisdictionAudit/useSetupMenuItems/_mocks'
 
 jest.unmock('react-toastify')
-jest.spyOn(utilities, 'api')
+
+const userMocks: { [key in 'ja' | 'aa']: IUserMeta } = {
+  ja: {
+    name: 'Han Solo',
+    email: 'falcon@gmail.com',
+    type: 'jurisdiction_admin',
+    organizations: [],
+    jurisdictions: [],
+  },
+  aa: {
+    name: 'Leia Organa',
+    email: 'princess@rebelalliance.com',
+    type: 'audit_admin',
+    organizations: [],
+    jurisdictions: [],
+  },
+}
+
+const apiMocks = {
+  failedAuth: {
+    url: '/api/me',
+    response: {},
+    error: {
+      status: 401,
+      statusText: 'UNAUTHORIZED',
+    },
+  },
+  successAuth: (response: IUserMeta) => ({
+    url: '/api/me',
+    response,
+  }),
+  abAuth: {
+    url: '/api/me',
+    response: { type: 'AUDIT_BOARD', ...dummyBoards()[1] },
+  },
+}
+
+const renderView = (route: string) => renderWithRouter(<App />, { route })
 
 describe('App', () => {
-  it('renders properly', async () => {
-    const { container } = render(<App />)
-    await screen.findByAltText('Arlo, by VotingWorks')
-    expect(container).toMatchSnapshot()
+  describe('/', () => {
+    it('renders unauthenticated properly', async () => {
+      const expectedCalls = [apiMocks.failedAuth]
+      await withMockFetch(expectedCalls, async () => {
+        const { container } = renderView('/')
+        await screen.findByAltText('Arlo, by VotingWorks')
+        expect(container).toMatchSnapshot()
```

Comment:
i have been finding these kinds of tests a bit difficult to maintain and review. I'm not totally sure exactly what they are testing for. the snapshots are so big that I can't really read them, so I don't know exactly what we're expecting (or at least what are the relevant things we're looking for). also it's a bit annoying because the snapshots change whenever we change styles (though that's a separate issue)

here are a few of my guesses for what these tests are about and suggestions for how they might be improved (but let me know if I'm misunderstanding):
- we're trying to test that it doesn't render empty - suggestion: remove the snapshots and just check for one relevant element to be rendered
- we're trying to test that it renders exactly the UI that we want - suggestion: check for the relevant pieces of the UI directly by finding them in the screen and using them - this would make it much easier to read the test and know what we're looking for
- we're trying to get code coverage up - not really sure what to do in this case, but might be worth thinking harder about

Reply:
@jonahkagan Good questions. Here are some of my thoughts.

The code coverage aspect doesn't change at all regardless of whether we use snapshots or not. As long as it renders, it's counted as covered.

However, testing each individual significant piece is far more verbose than just capturing it all as a snapshot. We aren't just testing specific pieces, we're testing to make sure nothing changes that we didn't expect. There's been many times where something changed in one part of the codebase that resulted in changes elsewhere that were unanticipated (and undesired) which was only noticed because the snapshots changed and sounded the alarm.

As far as readability goes, there are plugins in VS Code which can help with that I think. I haven't used them yet, though. Maybe they'd help?

Reply:
Whoops sorry I let this convo slip. Thanks for explaining - helpful to understand that it's not so much about coverage but more about trying to lock in the content of the UI.

Personally, I think being verbose and explicit would make for more useful and readable tests. But this might just be a matter of opinion and style, since it sounds like the snapshot tests have been useful for you.